pkg/internalregistry/controllers/keyid_observation_controller.go

```diff
@@ -0,0 +1,192 @@
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/keyutil"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+)
+
+type keyIDObservation struct {
+	client     kubernetes.Interface
+	secrets    listers.SecretLister
+	cacheSyncs []cache.InformerSynced
+	queue      workqueue.RateLimitingInterface
+	ch         kidsChan
+}
+
+func NewKeyIDObservationController(kubeclient kubernetes.Interface, secrets informers.SecretInformer, ch kidsChan) *keyIDObservation {
+	c := &keyIDObservation{
+		client:     kubeclient,
+		secrets:    secrets.Lister(),
+		ch:         ch,
+		cacheSyncs: []cache.InformerSynced{secrets.Informer().HasSynced},
+		queue:      workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "sa-signing-key-secrets"),
+	}
+	secrets.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj any) bool {
+			if secret, ok := obj.(*corev1.Secret); ok {
+				return (secret.Namespace == "openshift-kube-apiserver") && (secret.Name == "bound-service-account-signing-key")
+			}
+			return false
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj any) {
+				key, err := cache.MetaNamespaceKeyFunc(obj)
+				if err == nil {
+					c.queue.Add(key)
+				}
+			},
+			UpdateFunc: func(_, obj any) {
+				key, err := cache.MetaNamespaceKeyFunc(obj)
+				if err == nil {
+					c.queue.Add(key)
+				}
+			},
+		},
+	})
+	return c
+}
+
+func (c *keyIDObservation) sync(ctx context.Context, key string) error {
+	klog.V(4).InfoS("sync", "key", key)
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	secret, err := c.secrets.Secrets(ns).Get(name)
+	if errors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	pem, ok := secret.Data["service-account.pub"]
+	if !ok {
+		return fmt.Errorf("expected data service-account.pub not found")
+	}
+	keys, err := keyutil.ParsePublicKeysPEM(pem)
+	if err != nil {
+		return err
+	}
+	// compute key ID (e.g. hash) for token signing private keys from the public keys
+	var kids []string
+	for _, key := range keys {
+		der, err := x509.MarshalPKIXPublicKey(key)
+		if err != nil {
+			return err
+		}
+		hashFunc := crypto.SHA256.New()
+		hashFunc.Reset()
+		_, err = hashFunc.Write(der)
+		if err != nil {
+			return err
+		}
+		kids = append(kids, base64.RawURLEncoding.EncodeToString(hashFunc.Sum(nil)))
+	}
+	slices.Sort(kids)
+	select {
+	case c.ch <- kids:
+	case <-ctx.Done():
+	}
+	return nil
+}
+
+func (c *keyIDObservation) fallbackWorker(ctx context.Context) {
+	err := wait.PollUntilContextTimeout(ctx, 30*time.Second, 10*time.Minute, true, func(ctx context.Context) (bool, error) {
+		_, err := c.secrets.Secrets("openshift-kube-apiserver").Get("bound-service-account-signing-key")
+		if err != nil && !errors.IsNotFound(err) {
+			// signing key secret exists, skip and let sync handle
+			return true, nil
+		}
+		if err != nil {
+			runtime.HandleError(err)
+			return false, nil
+		}
```

Comment:
Does this ever need to return true (and stop)?

```suggestion
		if err == nil {
			// signing key secret exists, skip and let sync handle
			return false, nil
		} else if !errors.IsNotFound(err) {
			// error other than notfound, signing key secret may or may not exist
			runtime.HandleError(err)
			return false, nil
		}
```

Reply:
If the secret DNE, return `true` and stop the inner polling loop. `runWorker()` itself is running in an `UntilContext` loop.
If there is an error, the inner loop retries every 10 seconds for a minute.